scanner/src/main/kotlin/utils/FileListingResolver.kt

```diff
@@ -0,0 +1,106 @@
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.utils
+
+import java.io.File
+import java.util.LinkedList
+
+import org.apache.commons.compress.compressors.xz.XZCompressorInputStream
+import org.apache.commons.compress.compressors.xz.XZCompressorOutputStream
+
+import org.ossreviewtoolkit.model.HashAlgorithm
+import org.ossreviewtoolkit.model.KnownProvenance
+import org.ossreviewtoolkit.model.fromYaml
+import org.ossreviewtoolkit.model.toYaml
+import org.ossreviewtoolkit.model.utils.ProvenanceFileStorage
+import org.ossreviewtoolkit.scanner.FileListing
+import org.ossreviewtoolkit.scanner.provenance.ProvenanceDownloader
+import org.ossreviewtoolkit.utils.common.FileMatcher
+import org.ossreviewtoolkit.utils.common.VCS_DIRECTORIES
+import org.ossreviewtoolkit.utils.ort.createOrtTempFile
+
+internal class FileListingResolver(
+    private val storage: ProvenanceFileStorage,
+    private val provenanceDownloader: ProvenanceDownloader
+) {
+    fun resolve(provenance: KnownProvenance): FileListing {
+        storage.readFileListing(provenance)?.let { return it }
+
+        return provenanceDownloader.download(provenance).let { dir ->
+            createFileListing(dir).also { storage.writeFileListing(provenance, it) }
+        }
+    }
+}
+
+private fun ProvenanceFileStorage.writeFileListing(provenance: KnownProvenance, fileListing: FileListing) {
+    val tempFile = createOrtTempFile(prefix = "file-listing", suffix = ".yml.xz")
+
+    fileListing.toYaml().byteInputStream().use { input ->
+        XZCompressorOutputStream(tempFile.outputStream()).use { output ->
+            input.copyTo(output)
+        }
+    }
+
+    putFile(provenance, tempFile)
+    tempFile.delete()
+}
+
+private fun ProvenanceFileStorage.readFileListing(provenance: KnownProvenance): FileListing? {
+    val file = getFile(provenance) ?: return null
+
+    val yaml = XZCompressorInputStream(file.inputStream()).use { input ->
+        input.bufferedReader().readText()
+    }
+
+    return yaml.fromYaml()
+}
+
+private val IGNORED_DIRECTORY_MATCHER by lazy {
+    FileMatcher(patterns = VCS_DIRECTORIES.map { "**/$it" })
+}
+
+private fun createFileListing(dir: File): FileListing {
+    val files = mutableSetOf<FileListing.FileEntry>()
+    val queue = LinkedList<File>().apply { add(dir) }
+
+    while (queue.isNotEmpty()) {
+        queue.removeFirst().listFiles().forEach { child ->
+            val path = child.relativeTo(dir).invariantSeparatorsPath
+
+            when {
+                child.isFile -> files += FileListing.FileEntry(
+                    path = path,
+                    sha1 = HashAlgorithm.SHA1.calculate(child)
+                )
+                child.isDirectory -> if (!IGNORED_DIRECTORY_MATCHER.matches(path)) {
+                    queue += child
+                } else {
+                    println("ignoring: $path")
+                }
+                else -> return@forEach
+            }
+        }
+    }
```

Comment:
Why not simply use `File.walk()`instead of a custom queue implementation?

Reply:
I changed it to using `walk()`.